src/clients/c++/library/request_grpc.cc

```diff
@@ -938,61 +900,27 @@ void
 InferGrpcContextImpl::AsyncTransfer()
 {
   do {
-    // sleep if no work is available
-    std::unique_lock<std::mutex> lock(mutex_);
-    cv_.wait(lock, [this] {
-      if (this->exiting_) {
-        return true;
-      }
-      // wake up if at least one request is not ready
-      for (auto& ongoing_async_request : this->ongoing_async_requests_) {
-        if (std::static_pointer_cast<GrpcRequestImpl>(
-                ongoing_async_request.second)
-                ->IsReady() == false) {
-          return true;
-        }
-      }
-      return false;
-    });
-    lock.unlock();
     // GRPC async APIs are thread-safe https://github.com/grpc/grpc/issues/4486
-    if (!exiting_) {
-      std::shared_ptr<Request> request_with_callback;
-      size_t got;
-      bool ok = true;
-      bool status = async_request_completion_queue_.Next((void**)(&got), &ok);
-      {
-        std::lock_guard<std::mutex> lock(mutex_);
-        if (!ok) {
-          fprintf(stderr, "Unexpected not ok on client side.");
-        }
-        if (!status) {
-          fprintf(stderr, "Completion queue is closed.");
-        }
-        auto itr = ongoing_async_requests_.find(got);
-        if (itr == ongoing_async_requests_.end()) {
-          fprintf(
-              stderr,
-              "Unexpected error: received completed request that"
-              " is not in the list of asynchronous requests.\n");
-          continue;
-        }
-
-        std::shared_ptr<GrpcRequestImpl> grpc_request =
-            std::static_pointer_cast<GrpcRequestImpl>(itr->second);
-        grpc_request->Timer().CaptureTimestamp(
-            RequestTimers::Kind::REQUEST_END);
-        grpc_request->SetIsReady(true);
-        if (grpc_request->HasCallback()) {
-          request_with_callback = itr->second;
-        }
+    GrpcRequestImpl* grpc_request_ptr;
+    bool ok = true;
+    bool status =
+        async_request_completion_queue_.Next((void**)(&grpc_request_ptr), &ok);
+    if (!ok) {
+      fprintf(stderr, "Unexpected not ok on client side.\n");
+    }
+    if (!status) {
+      if (!exiting_) {
+        fprintf(stderr, "Completion queue is closed.\n");
       }
-      // send signal in case the main thread is waiting
-      cv_.notify_all();
-      if (request_with_callback != nullptr) {
-        GrpcRequestImpl* request =
-            static_cast<GrpcRequestImpl*>(request_with_callback.get());
-        request->callback_(this, request_with_callback);
+    } else if (grpc_request_ptr == nullptr) {
+       fprintf(stderr, "Unexpected null tag received at client.\n");
+      } else {
+      grpc_request_ptr->Timer().CaptureTimestamp(
```

Comment:
You should put grpc_request_ptr into a shared_ptr as soon as you know it is not nullptr (in fact you could do it immeidately after the Next() call above, it doesn't matter if it is null or not. As it is now you have a leak when the callback_ is nullptr.